x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/allocation/decider/SearchableSnapshotRepositoryExistsAllocationDecider.java

```diff
@@ -0,0 +1,101 @@
+import org.elasticsearch.cluster.routing.allocation.decider.Decision;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshotsConstants;
+
+import java.util.List;
+
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_REPOSITORY_NAME_SETTING;
+import static org.elasticsearch.xpack.searchablesnapshots.SearchableSnapshots.SNAPSHOT_REPOSITORY_UUID_SETTING;
+
+public class SearchableSnapshotRepositoryExistsAllocationDecider extends AllocationDecider {
+
+    private static final String NAME = "searchable_snapshot_repository_exists";
+
+    private static final Decision YES_INAPPLICABLE = Decision.single(
+        Decision.Type.YES,
+        NAME,
+        "this decider only applies to indices backed by searchable snapshots"
+    );
+
+    private static final Decision YES_REPOSITORY_EXISTS = Decision.single(
+        Decision.Type.YES,
+        NAME,
+        "the repository containing the data for this index exists"
+    );
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return allowAllocation(allocation.metadata().getIndexSafe(shardRouting.index()), allocation);
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingAllocation allocation) {
+        return allowAllocation(allocation.metadata().getIndexSafe(shardRouting.index()), allocation);
+    }
+
+    @Override
+    public Decision canAllocate(IndexMetadata indexMetadata, RoutingNode node, RoutingAllocation allocation) {
+        return allowAllocation(indexMetadata, allocation);
+    }
+
+    private static Decision allowAllocation(IndexMetadata indexMetadata, RoutingAllocation allocation) {
+        final Settings settings = indexMetadata.getSettings();
+        if (SearchableSnapshotsConstants.isSearchableSnapshotStore(settings)) {
+
+            final RepositoriesMetadata repositoriesMetadata = allocation.metadata().custom(RepositoriesMetadata.TYPE);
+            if (repositoriesMetadata == null || repositoriesMetadata.repositories().isEmpty()) {
+                return allocation.decision(Decision.NO, NAME, "there are no repositories registered in this cluster");
+            }
+
+            final String repositoryUuid = SNAPSHOT_REPOSITORY_UUID_SETTING.get(settings);
+            if (Strings.hasLength(repositoryUuid)) {
+                final List<RepositoryMetadata> repositories = repositoriesMetadata.repositories();
+                if (repositories.stream().anyMatch(r -> repositoryUuid.equals(r.uuid()))) {
+                    return YES_REPOSITORY_EXISTS;
+                }
+
+                return allocation.decision(
+                    Decision.NO,
+                    NAME,
+                    "this index is backed by a searchable snapshot in a repository with UUID [%s] but no such repository is registered "
```

Comment:
I wonder if we should add more to the message if a repo exists with the right name but wrong UUID?

Reply:
I can see some value in that but I'm inclined not to change it until we see evidence that it's needed. We expose the repository UUIDs in various APIs, e.g. `GET _snapshot` so I think this message should be enough to point the user in the right direction in most cases.